content/client-lib-development-guide/features.textile

```diff
@@ -950,19 +950,23 @@ h2(#push-notifications). Push notifications
 ** @(RSH2b)@ @Push#deactivate@ sends a @CalledDeactivate@ event to "the state machine":#RSH3.
-** @(RSH2c)@ Whenever any change arises of the push transport details for local device (eg an FCM registration token update triggered by the platform), a @GotPushDeviceDetails@ event is sent to "the state machine":#RSH3.
-** @(RSH2d)@ If an attempt to obtain the push transport details for local device (eg an FCM registration token) fails, a @GettingPushDeviceDetailsFailed@ event containing the indicated error is sent to "the state machine":#RSH3.
-** @(RSH2e)@ Each time the library is instanced, if the LocalDevice has push device details (eg an APNS deviceToken), it must verify the validity of those details (eg by requesting a token from the platform and comparing that with the already-known token). If as a result there are updated details, then an update to the Ably server is triggered by sending a @GotPushDeviceDetails@ event to "the state machine":#RSH3.
+** @(RSH2c)@ (Moved to @(RSH8g)@).
+** @(RSH2d)@ (Moved to @(RSH8h)@).
+** @(RSH2e)@ (Moved to @(RSH8i)@).
 
-h3(#activation-state-machine). Activation state machine
+h3(#activation-state-machine). Activation State Machine
 
 * @(RSH3)@ In platforms that support receiving push notifications, in order to connect the device's push features with Ably's, the library must perform the process described in the following abstract state machine. While this process should be implemented in whatever way better fits the concrete platform, it should be taken into account that its lifetime is that of the _app_ that runs it, which outlives that of the @Rest@ instance or (typically) the process running the app. This typically forces some kind of on-disk storage to which the state machine's state must be persisted, so that it can be recovered later by new instances and processes running the app triggered by external events.
 ** @(RSH3a)@ State @NotActivated@ (the initial one).
 *** @(RSH3a1)@ On event @CalledDeactivate@:
 **** @(RSH3a1a)@ Makes @Push#deactivate@ return or call its callback with no error.
 **** @(RSH3a1b)@ Transitions to @NotActivated@.
 *** @(RSH3a2)@ On event @CalledActivate@:
-**** @(RSH3a2a)@ If the local device has @deviceIdentityToken@, enqueues a @CalledActivate@ event and transitions to @WaitingForNewPushDeviceDetails@ and #RSH3a2b onwards don't apply.
+**** @(RSH3a2a)@ If the local device has @deviceIdentityToken@, enqueues a @CalledActivate@ event and performs a validation of the local DeviceDetails via the following steps. #RSH3a2b onwards then don't apply.
+***** @(RSH3a2a1)@ Checks the compatibilty of the present client with the existing registration: if the @LocalDevice@ has a non-empty @clientId@, and the present identified client has a different (non-null) @clientId@, then a @SyncRegistrationFailed@ event should be fired containing an error with @code@ 61002, and no further validation steps performed.
```

Comment:
On implementing this, I'm realizing it's unclear whether RSH3a2a4 applies after RSH3a2a1 results in a failure.

It _should_ apply, since WaitingForRegistrationSync is the state that has the handler for the events that RSH3a2a4 (and any RSH3a2a*) may emit, but the hierarchy makes it seem like it doesn't.

I think RSH3a2a4 is just RSH3a2e; it's not a validation, it's the state transition we do no matter the side effects we start. So the hierarchy should reflect that, more or less like this:

* State `NotActivated`
  * On event `CalledActivate`
    * RSH3a2a: if we have `deviceIdentityToken`, performs the following:
       * client ID validation, may emit `SyncRegistrationFailed`
       * PUT or custom registrar, may emit `SyncRegistrationFailed` or `RegistrationSynced`
    * Otherwise:
      * RSH3a2b
      * RSH3a2c
      * RSH3a2d
    * Transitions to WaitingForPushDeviceDetails

Reply:
Sorry, that's totally wrong, RSH3a2a4 is not just RSH3a2e, they transition to different states (`WaitingForPushDeviceDetails` and `WaitingForPushDeviceDetails`).

Still, "no further validation steps performed" should maybe just be "skips until RSH3a2a4".

Reply:
Done at 40532aa64b6db94e59f474002e349f3d8e8396a5.